### PR TITLE
Challenge persistence

### PR DESCRIPTION
**Pull request by rjacobs31**
- [X] Add challenge persistence on challenge manager.
- [X] Fetch persisted challenge on challenge activity load.
- [X] Clear challenge data on exceptions.
- [X] Clear active challenge data if expired.

Other:
- [X] Add parameterized map type.
- [X] Add type-based complex-object getter in shared preferences.
- [X] Remove old module.

### CODE REVIEW

**Comment by rjacobs31** [comment]
Hi, @wvictor-rr. Would you please take a look at this PR?

**Comment by rjacobs31** [comment]
Minor bug with the registration button text always becoming "Continue". Fixing.

**Comment by rjacobs31** [comment]
Fixed the bug.

File: app/src/main/java/org/gem/indo/dooit/views/main/fragments/challenge/ChallengeFragment.java [hunk 1]
```diff
@@ -94,40 +92,55 @@ public void onDestroyView() {
...
+                Log.d(TAG, "Could not load persisted challenge");
+                persisted.clearCurrentChallenge();
+            }
+        }
+
         challengeSubscription = challengeManager.retrieveCurrentChallenge(false, new DooitErrorHandler() {
             @Override
             public void onError(DooitAPIError error) {
-                Toast.makeText(getContext(), "Could not retrieve challenges", Toast.LENGTH_SHORT).show();
+                Toast.makeText(getContext(), "Could not retrieve challenge", Toast.LENGTH_SHORT).show();
```
Comment (wvictor-rr) [inline app/src/main/java/org/gem/indo/dooit/views/main/fragments/challenge/ChallengeFragment.java:135]: Put toast string in strings.xml

**Review by wvictor-rr — APPROVED** [review]

**Pull request merged**